src/test/java/org/folio/services/migration/BatchedReadStreamTest.java

```diff
@@ -0,0 +1,173 @@
+package org.folio.services.migration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.folio.rest.api.TestBase.get;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.streams.ReadStream;
+import io.vertx.sqlclient.Row;
+import java.util.HashSet;
+import java.util.List;
+import java.util.UUID;
+import org.folio.rest.api.ReindexJobRunnerTest;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+public class BatchedReadStreamTest {
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldProcessRecordsByBatch() {
+    var numberOfRecords = 410;
+    var batchSize = 100;
+    var numberOfBatches = 5;
+    var rowStream = new ReindexJobRunnerTest.TestRowStream(numberOfRecords);
+    var batchedReadStream = new BatchedReadStream<>(rowStream, batchSize);
+    var promise = Promise.promise();
+    Handler<List<Row>> handler = mock(Handler.class);
+
+    batchedReadStream
+      .handler(handler)
+      .endHandler(promise::tryComplete)
+      .exceptionHandler(promise::tryFail);
+
+    // Wait until completed
+    get(promise.future());
+
+    ArgumentCaptor<List<Row>> captor = ArgumentCaptor.forClass(List.class);
+    verify(handler, times(numberOfBatches)).handle(captor.capture());
+
+    assertThat(captor.getAllValues()).hasSize(numberOfBatches);
+    assertThat(captor.getAllValues().get(4)).hasSize(10);
+    captor.getAllValues().stream().limit(numberOfBatches - 1)
+      .forEach(list -> assertThat(list).hasSize(batchSize));
+
+
+    var ids = new HashSet<UUID>(numberOfRecords);
+    captor.getAllValues().forEach(
+      list -> list.forEach(
+        // Make sure all IDS are unique so that batch doe not produce
+        // duplicate records
+        row -> assertThat(ids.add(row.getUUID("id"))).isTrue()));
+  }
+
+  @Test
+  public void shouldDoNothingIfNullHandler() {
+    var rowStream = mockStream();
+    var batchedReadStream = new BatchedReadStream<>(rowStream);
+
+    batchedReadStream.handler(null);
+
+    verifyNoInteractions(rowStream);
+  }
+
+  @Test
+  public void shouldDelegateCallForExceptionHandler() {
+    var delegate = mockStream();
```

Comment:
Does this mean that if a the exception handler is invoked for the `BatchReadStream` it invokes the exception handler for the `ReadStream`?